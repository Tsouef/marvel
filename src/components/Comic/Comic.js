import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchComicById } from '../../actions';
import './Comic.css';

class Comic extends Component {
  componentDidMount() {
    this.props.fetchComicById(this.props.match.params.id);
  }

  renderCharacters(characters) {
    return characters.map(character => {
      return <li key={character.name}>{character.name}</li>;
    });
  }

  render() {
    const {
      thumbnail,
      title,
      description,
      prices,
      characters
    } = this.props.comic;

    if (this.props.comic.length <= 0) {
      return <div>Waiting</div>;
    }

    return (
      <div className="comic">
        <div className="comic-cover">
          <img alt="" src={`${thumbnail.path}.${thumbnail.extension}`} />
        </div>
        <div className="comic-description">
          <p>
            <span className="meta">Title: </span> {title}
          </p>
          <p>
            <span className="meta">Description: </span>
            <span dangerouslySetInnerHTML={{ __html: description }} />
          </p>
          <p>
            <span className="meta">Price: </span> {prices[0].price}$
          </p>
          <p>
            <span className="meta">Characters: </span>
            <ul className="comic-characters">
              {this.renderCharacters(characters.items)}
            </ul>
          </p>
        </div>
      </div>
    );
  }
}

function mapStateToProps({ comic }) {
  return { comic };
}

export default connect(mapStateToProps, { fetchComicById })(Comic);
