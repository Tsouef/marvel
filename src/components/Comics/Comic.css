.comic {
  margin: 20px 0 0 0;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  flex-wrap: nowrap;
}

.comic-description {
  display: flex;
  flex-direction: column;
  margin: 0 0 0 20px;
}

.meta {
  font-weight: bold;
  font-size: 15px;
}

.comic-characters li {
  margin: 0 0 0 15px;
  line-height: 1.4;
}
