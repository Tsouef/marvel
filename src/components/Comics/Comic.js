import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchComicById } from '../../actions';
import './Comic.css';

export class Comic extends Component {
  componentDidMount() {
    if (this.props.match) {
      this.props.fetchComicById(this.props.match.params.id);
    }
  }

  renderCharacters(characters) {
    return characters.items.map(character => {
      return <li key={character.name}>{character.name}</li>;
    });
  }

  render() {
    if (this.props.comics.isFetching || this.props.comics.comic.length <= 0) {
      return <div>Loading...</div>;
    }
    const {
      thumbnail,
      title,
      description,
      prices,
      characters
    } = this.props.comics.comic;

    return (
      <div className="comic">
        <div className="comic-cover">
          <img alt="" src={`${thumbnail.path}.${thumbnail.extension}`} />
        </div>
        <div className="comic-description">
          <p>
            <span className="meta">Title: </span> {title}
          </p>
          <p>
            <span className="meta">Description: </span>
            <span dangerouslySetInnerHTML={{ __html: description }} />
          </p>
          <p>
            <span className="meta">Price: </span> {prices[0].price}$
          </p>
          <p>
            <span className="meta">Characters: </span>
            <ul className="comic-comics">
              {this.renderCharacters(characters)}
            </ul>
          </p>
        </div>
      </div>
    );
  }
}

function mapStateToProps({ comics }) {
  return { comics };
}

export default connect(mapStateToProps, { fetchComicById })(Comic);
