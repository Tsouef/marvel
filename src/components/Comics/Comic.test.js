import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Comic } from './Comic';

describe('Comic Component', () => {
  it('should render Comic', () => {
    const props = {
      comics: {
        comic: [],
        isFetching: false
      }
    };

    const tree = renderer
      .create(
        <MemoryRouter>
          <Comic {...props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with props', () => {
    const props = {
      comics: {
        comic: {
          thumbnail: { path: 'path', extension: 'jpg' },
          title: 'comic',
          prices: [{ type: 'printPrice', price: 0 }],
          description: 'text',
          characters: {
            items: [
              {
                resourceURI:
                  'http://gateway.marvel.com/v1/public/characters/1009220',
                name: 'Captain America'
              }
            ]
          }
        },
        isFetching: false
      }
    };

    const tree = renderer
      .create(
        <MemoryRouter>
          <Comic {...props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
