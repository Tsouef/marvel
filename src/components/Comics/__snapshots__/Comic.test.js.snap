// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comic Component should render Comic 1`] = `
<div>
  Loading...
</div>
`;

exports[`Comic Component should render with props 1`] = `
<div
  className="comic"
>
  <div
    className="comic-cover"
  >
    <img
      alt=""
      src="path.jpg"
    />
  </div>
  <div
    className="comic-description"
  >
    <p>
      <span
        className="meta"
      >
        Title: 
      </span>
       
      comic
    </p>
    <p>
      <span
        className="meta"
      >
        Description: 
      </span>
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "text",
          }
        }
      />
    </p>
    <p>
      <span
        className="meta"
      >
        Price: 
      </span>
       
      0
      $
    </p>
    <p>
      <span
        className="meta"
      >
        Characters: 
      </span>
      <ul
        className="comic-comics"
      >
        <li>
          Captain America
        </li>
      </ul>
    </p>
  </div>
</div>
`;
